pyro/autohelp/autohelp.py

```diff
@@ -128,254 +153,182 @@ async def process_message(
         if key in self._help_cache:
             return None
 
-        code_bin_content = await self._code_bin.process(message.content)
-        message.content += code_bin_content
-        message.content = message.content = message.content.replace("\r", "")
+        contents = await self.find_code(message)
+        if not contents:
+            return
 
-        iters = [call(message) for call in self.patterns]
-        results = await asyncio.gather(*iters)
-        results = list(filter(None, results))
-        if not results:
+        # parse the contents
+        # test the first valid code snippet for now
+        total_old = ''
+        total_new = ''
+        errors = []
+        # attempt to parse the full code
+        for num , code in enumerate(['\n'.join(contents),*contents]):
+            try:
+                parsed = libcst.parse_module(code)
+            except libcst.ParserSyntaxError:
+                continue
+            results = parsed
+            total_old += parsed.code + '\n'
+            visitors: list[libcst.CSTTransformer] = []
+            local_errors = []
+            for visitor in self.visitors:
+                visitor = visitor()
+                visitors.append(visitor)
+                results = results.visit(visitor)
+                if visitor.found_errors:
+                    local_errors.extend(visitor.found_errors)
+
+            total_new += results.code + '\n'
+            if local_errors:
+                errors.extend(local_errors)
+
+            if num == 0:
+                # stop short if we're running all code blocks successfully
+                break
+
+        if not errors:
             return None
 
-        self._help_cache.add_entry(key, None, ttl=datetime.timedelta(minutes=30))
```

Comment:
This needs to be added back. However ``key`` needs to be changed to also take into account the actions taken.

This also should probably be moved to after parsing stuffs